polygeist/tools/cgeist/driver.cc

```diff
@@ -361,13 +361,17 @@ static int optimize(mlir::MLIRContext &Ctx,
   mlir::PassManager PM(&Ctx);
   enableOptionsPM(PM);
 
-  mlir::OpPassManager &OptPM = PM.nestAny();
   GreedyRewriteConfig CanonicalizerConfig;
   CanonicalizerConfig.maxIterations = CanonicalizeIterations;
 
   if (OptLevel != llvm::OptimizationLevel::O0) {
     PM.addPass(polygeist::createArgumentPromotionPass());
 
+    mlir::OpPassManager &OptPM = PM.nestAny();
```

Comment:
LICM needs to be after ArgumentPromotion, in order to catch the matrix mulitply reduction opportunity.